usr/share/rear/lib/rsync-functions.sh

```diff
@@ -74,25 +74,39 @@ function rsync_path () {
     local url="$1"
     local host
     local path
+    local url_without_scheme
+    local url_without_scheme_user
 
     host=$(url_host "$url")
     path=$(url_path "$url")
     local tmp2="${host#*@}"
 
+    url_without_scheme="${url#*//}"
+    url_without_scheme_user="${url_without_scheme#$(rsync_user "$url")@}"
```

Comment:
@pcahyna 
only a question out of curiosity:

For now all is OK i.e. no changes for ReaR 2.7
but I like to understand if something may have
to be done for ReaR 2.8 e.g. further cleanup.

I wonmder what the reason is why you cannot use
the url_username() function in lib/global-functions.sh
to get the user name?

Your rsync_user() function parses what url_host() results
but I made url_username() and url_password() functions which
should work generically for URIs/URLs which are of the form
```
scheme:[//[user:password@]host[:port]][/]path[?query][#fragment]
```

I guess for rsync URLs things are special so the generic
url_username() and url_password() functions don't work.

Or is there perhaps something wrong with the
current url_username() and url_password() functions
that should be enhanced of fixed for ReaR 2.8?

Reply:
Short answer: I don't know, my code originated as a copy of the original rsync URL parsing stuff in 
https://github.com/rear/rear/blob/a865e33e12d55f3f92ec06eda0efb147c153076b/usr/share/rear/prep/RSYNC/default/100_check_rsync.sh . I though it is a good idea to only move code around and make it more generic without changing how it works, because it reduces the likelihood of regressions. It is well possible that the generic URL functions are well sufficient for this purpose. (Note though that not all URL functions are suitable for all cases in this file, because they don't cope well with URLs where the path is not separated by a slash from host, as it is the case with old-style rsync URLs, where `:` or `::` is used as the separator.) It would take more analysis and experimenting to determine whether `url_username()` falls in the "suitable" or in the "unsuitable" category, which can be done for 2.8 indeed. (Note: I am not at all sure that rsync URLs support an embedded password, especially if `:` is used as a separator between user and password, since `:` is already a bit overloaded - it separates host from path in old-style scp-like rsync-over-ssh URLs, and also host from port in rsync protocol URLs.)

Reply:
I don't believe that the syntax `scheme:[//[user:password@]host[:port]][/]path[?query][#fragment]` is entirely accurate. At least the `/` separating host and path does not seem to be optional for our `url_` functions.

Reply:
Via
https://github.com/rear/rear/commit/606cb40607b1a90b91e44c8744c49c3e53adb203
I fixed the comment in lib/global-functions.sh
that shows a generic URI/URL syntax so it shows now
```
scheme:[//[[user:password@]host[:port]]/path[?query][#fragment]
```

Reply:
Via
https://github.com/rear/rear/commit/bdc467363e5189cecd4a649de7947eb59b064fa3
I fixed the before falsely fixed comment
that shows now only a usual generic URI/URL syntax so it is now
```
scheme://[[user:password@]host[:port]]/path[?query][#fragment]
```
so the case `scheme:path` (like `mailto:John.Doe@example.com`)
is left out for simplicity in that comment because
I don't know if the current `url_...` functions
support the `scheme:path` case.

Reply:
They don't, actually. `url_scheme()` searches for `://`, not `:` (which makes sense, as `:` is used in old-style rsync "URLs" (that are not real URLs) as a separator). So your change is correct. `://` is much more visually clear anyway, so I would not even attempt to support `scheme:path`, unless there is a real need.

Reply:
Yes, that the old style rsync "URLs" are no URLs
(because `[USER@]HOST:[:]PATH` is invalid URI syntax)
is another good reason to clean that up properly
when there is real need for such a cleanup, cf.
https://github.com/rear/rear/issues/2781#issuecomment-1181617916